robot_smach_states/src/robot_smach_states/util/designators/arm.py

```diff
@@ -13,6 +15,24 @@
 __author__ = 'loy'
 
 
+def const_resolve(arm_designator, additional_properties):
```

Comment:
You might consider calling this something like `resolve_extended`. I think that the `resolve` should always be const. This method is useful to resolve a designator with additional demands without changing the original. I'm curious about @LoyVanBeek 's view on this matter

Reply:
I think this is not the way to do things. If a state requires a more specific arm than other states, that `ArmDesignator`, shared with those other states, should be extended. Otherwise, for example, state A has different, more lax requirements and ends up with a different arm than the more demanding state B, which are both part of the same process that should use the *same* arm in the end. 

Extending the arm requirements is OK, cloning designators to alter and potentially have the resolve differently is not.

Reply:
Thanks for your feedback. I guess that depends on where you define the requirements and how you implement it.

The usecase is thus the `ArmDesignator`. Suppose I build a hierarchical state machine: from the top level, I don't want to worry about the requirements I need to add. However, I do want 

1. a check at startup to make sure my robot has actually the required hardware to complete the task
2. at runtime, when I I resolve my designator, that my arm still meets the same requirement.

In the current implementation (the one in master), I need to add *all* requirements to the top-level designator. Even in case of a simple example, this is extremely unpractical (see, e.g., the arm designator in [this script](https://github.com/tue-robotics/tue_robocup/pull/1013/files#diff-1112c2269344b370678c16ef86befdfb).

The proposed implementation makes it therefore much easier for the user. If the user (at any level) wants to make sure a designator resolves to the same arm, it should be locked (that's what this method is for!). 

I actually consider it very bad practice if states are able to modify a designator (other than the `writing` that has been implemented). That makes it extremely hard to debug stuff: if you test a certain part of your code stand-alone it will behave different from a situation where it's integrated. 

This discussion, however, does bring up a flaw in the proposed design: what if a robot has two arms; one of them has a gripper, the other one knows what a 'prepare_grasp' trajectory is. In that case, it will pass the startup check but fail on runtime anyway. To prevent this, the level at which the arm check is performed should probably be the same as the level where it should be locked. For example, the arm designator in the `Grab` state should be locked throughout the entire state (otherwise, we'd better use a different method). Correspondingly, the `Grab` state should collect the requirements of all its child states, combine these and then perform the startup check.

Reply:
The locking is different than what you mention. That applies when e.g. I need an Entity on-top-of the table and I want to resolve the same ID after than object has been grasped off the table. That;s locking. 

In the case discussed here, I do think that modifying/extending/refining a designator (appending more requirements) is actually appropriate. By the end of the `machine.__init__` you know what arm  you really really want. And yes, if you use different states, you get different requirements. 

The order (the hierachy is a tree but the code of constructing that tree is still sequential) of when the requirements are added should not matter and I don't think it does given it's all dictionaries. 

We're in a pickle if things conflict though, but we can check for that. Eg. one state needs a suction cup and the other a parallel gripper etc.

Reply:
'The locking is different': in what sense? If I position my two-armed robot to grasp using the left arm I definitely want to use the same arm for the next actions. I actually once implemented this by passing userdata and last time we used AMIGO we actually found out that the current implementation of the grab state *actually does not work anymore*. This indicates we *should* have a decent implementation that can either be locking or resolving just once and subsequently using a different mechanism.

Furthermore, I'm still not convinced on your argument on mutability: it still leads to different behavior while the same inputs are provided and as mentioned we're mixing the locking and arm requirements issues.

Maybe we should discuss this face to face to come to an 'optimal' design.

Reply:
> Maybe we should discuss this face to face to come to an 'optimal' design.

Yes, definitely. 

Locking is different is that the *same* designator can resolve to different things over time, potentially. Locking solves that, same value until it's unlocked. 

The issue here is that, under the hood, we copy a designator and suddenly we have multiple designators, that can each resolve to a different thing at the same moment. For the arms, we want all the states using something to resolve to the same arm as well. 
If we lock it but has the wrong value because of different requirements, locking is not the solution.